src/_avif.c

```diff
@@ -0,0 +1,1084 @@
+
+    Py_BEGIN_ALLOW_THREADS result = avifImageRGBToYUV(frame, &rgb);
+    Py_END_ALLOW_THREADS
+
+        if (result != AVIF_RESULT_OK) {
+        PyErr_Format(
+            exc_type_for_avif_result(result),
+            "Conversion to YUV failed: %s",
+            avifResultToString(result)
+        );
+        ret = NULL;
+        goto end;
+    }
+
+    uint32_t addImageFlags = AVIF_ADD_IMAGE_FLAG_NONE;
+    if (PyObject_IsTrue(is_single_frame)) {
+        addImageFlags |= AVIF_ADD_IMAGE_FLAG_SINGLE;
+    }
+
+    Py_BEGIN_ALLOW_THREADS result =
+        avifEncoderAddImage(encoder, frame, duration, addImageFlags);
+    Py_END_ALLOW_THREADS
+
+        if (result != AVIF_RESULT_OK) {
+        PyErr_Format(
+            exc_type_for_avif_result(result),
+            "Failed to encode image: %s",
+            avifResultToString(result)
+        );
+        ret = NULL;
+        goto end;
+    }
+
+end:
+    avifRGBImageFreePixels(&rgb);
+    if (!is_first_frame) {
+        avifImageDestroy(frame);
+    }
+
+    if (ret == Py_None) {
+        self->frame_index++;
+        Py_RETURN_NONE;
+    } else {
+        return ret;
+    }
+}
+
+PyObject *
+_encoder_finish(AvifEncoderObject *self) {
+    avifEncoder *encoder = self->encoder;
+
+    avifRWData raw = AVIF_DATA_EMPTY;
+    avifResult result;
+    PyObject *ret = NULL;
+
+    Py_BEGIN_ALLOW_THREADS result = avifEncoderFinish(encoder, &raw);
+    Py_END_ALLOW_THREADS
+
+        if (result != AVIF_RESULT_OK) {
+        PyErr_Format(
+            exc_type_for_avif_result(result),
+            "Failed to finish encoding: %s",
+            avifResultToString(result)
+        );
+        avifRWDataFree(&raw);
+        return NULL;
+    }
+
+    ret = PyBytes_FromStringAndSize((char *)raw.data, raw.size);
+
+    avifRWDataFree(&raw);
+
+    return ret;
+}
+
+// Decoder functions
+PyObject *
+AvifDecoderNew(PyObject *self_, PyObject *args) {
+    PyObject *avif_bytes;
+    AvifDecoderObject *self = NULL;
+
+    char *upsampling_str;
+    char *codec_str;
+    avifCodecChoice codec;
+    avifChromaUpsampling upsampling;
+    int max_threads = 0;
+
+    avifResult result;
+
+    if (!PyArg_ParseTuple(
+            args, "Sssi", &avif_bytes, &codec_str, &upsampling_str, &max_threads
+        )) {
+        return NULL;
+    }
+
+    if (!strcmp(upsampling_str, "auto")) {
+        upsampling = AVIF_CHROMA_UPSAMPLING_AUTOMATIC;
+    } else if (!strcmp(upsampling_str, "fastest")) {
+        upsampling = AVIF_CHROMA_UPSAMPLING_FASTEST;
+    } else if (!strcmp(upsampling_str, "best")) {
+        upsampling = AVIF_CHROMA_UPSAMPLING_BEST_QUALITY;
+    } else if (!strcmp(upsampling_str, "nearest")) {
+        upsampling = AVIF_CHROMA_UPSAMPLING_NEAREST;
+    } else if (!strcmp(upsampling_str, "bilinear")) {
+        upsampling = AVIF_CHROMA_UPSAMPLING_BILINEAR;
+    } else {
+        PyErr_Format(PyExc_ValueError, "Invalid upsampling option: %s", upsampling_str);
+        return NULL;
+    }
```

Comment:
As GitHub Actions says above,
> variable ‘upsampling’ set but not used [-Wunused-but-set-variable]

So either this partially implemented feature should be removed, or it needs to finish being implemented?